source/renderer/app/components/widgets/forms/PinCode.js

```diff
@@ -31,42 +39,258 @@ export default class PinCode extends Component<Props> {
   inputsRef = [];
   focusKey = 0;
   add = false;
+  fromBackspace = false;
+
+  state = {
+    isBackSpace: false,
+    focusKeyChanged: false,
+    focusIsUpdated: false,
+  };
+
+  hasInputNewValue = (inputNewValue: string, key: number) => {
+    const { value } = this.props;
+    const valueIsEmpty =
+      !value[key] || value[key] === '' || inputNewValue === '';
+    const valueHasChanged =
+      inputNewValue &&
+      inputNewValue.length === 1 &&
+      value[key] !== inputNewValue;
+    return valueIsEmpty || valueHasChanged;
+  };
+
+  isInputValueSelected = (input: Field) => {
+    const { inputElement } = input;
+    const { current } = inputElement;
+    return current.selectionStart !== current.selectionEnd;
+  };
+
+  isInputFieldDisabled = (
+    index: number,
+    value: Array<string>,
+    disabled: boolean
+  ) => {
+    return (
+      disabled ||
+      (index !== 0 &&
+        (!value ||
+          (!value.length && !value[index - 1] && value[index] && !!value[this.focusKey]) ||
+          (value.length &&
+            index > this.focusKey &&
+            !value[index] &&
+            !value[this.focusKey])))
+    );
+  };
 
-  onChange = (inputValue: ?number, key: number) => {
+  onChange = (inputValue: ?number | ?string, key: number) => {
     const { value, onChange } = this.props;
+    const { isBackSpace } = this.state;
     const inputNewValue =
-      inputValue !== null && inputValue !== undefined
+      inputValue !== null && inputValue !== undefined && !isNaN(inputValue)
         ? inputValue.toString()
         : '';
-
-    if (
-      !Object.prototype.hasOwnProperty.call(value, key) ||
-      value[key] === '' ||
-      inputNewValue === ''
-    ) {
+    if (this.hasInputNewValue(inputNewValue, key)) {
       const newValue = value;
-      newValue[key] = inputNewValue;
+      if (!isNaN(inputValue)) {
+        // Recheck if user pressed backspace and moved cursor to previous input field which has value
+        if (
+          isBackSpace &&
+          newValue[key] !== '' &&
+          inputNewValue === '' &&
+          this.focusKey !== key
+        ) {
+          // Set old value to field because it was previously deleted while shifting focus
+          newValue[key] = value[key];
+          // Calculate new field key for focus
+          const focusKey =
+            this.inputsRef[key] &&
+            this.inputsRef[key].inputElement.current.selectionStart === 0 &&
+            key > 2 &&
+            key < this.inputsRef.length
+              ? key - 1
+              : key;
+          // Delay focus to the same field while waiting for the validation to pass
+          setTimeout(() => {
+            const inputFieldRef = this.inputsRef[focusKey];
+            if (inputFieldRef && inputFieldRef.inputElement) {
+              this.setFocusOnField(inputFieldRef);
+              this.setState({ focusKeyChanged: false, focusIsUpdated: true });
+            }
+          }, 0);
```

Comment:
Do we need timeout here because it is set to `0ms`

Reply:
This part handles switching of focus on the field after value is deleted from the input. We need to make sure its executed only in a specific case after backspace is clicked and all the logic related to it is already done with calculations where to move the focus, hence, setTimeout.